types/src/trusted_state.rs

```diff
@@ -0,0 +1,232 @@
+    /// The current verifier. If we're starting up fresh, this is probably a
+    /// waypoint from our config. Otherwise, this is probably generated from the
+    /// validator set in the last known epoch change ledger info.
+    verifier: VerifierType,
+}
+
+/// `TrustedStateChange` is the result of attempting to ratchet to a new trusted
+/// state. In order to reduce redundant error checking, `TrustedStateChange` also
+/// contains references to relevant items used to ratchet us.
+#[derive(Clone, Debug)]
+pub enum TrustedStateChange<'a> {
+    /// We have a newer `TrustedState` but it's still in the same epoch, so only
+    /// the latest trusted version changed.
+    Version {
+        new_state: TrustedState,
+        latest_li: &'a LedgerInfoWithSignatures,
+    },
+    /// We have a newer `TrustedState` and there was at least one epoch change,
+    /// so we have a newer trusted version and a newer trusted validator set.
+    Epoch {
+        new_state: TrustedState,
+        latest_li: &'a LedgerInfoWithSignatures,
+        latest_epoch_change_li: &'a LedgerInfoWithSignatures,
+        latest_validator_set: &'a ValidatorSet,
+    },
+    /// Our current `TrustedState` is more recent.
+    Stale,
+}
+
+impl TrustedState {
+    /// Create an initial trusted state from a waypoint.
+    pub fn from_waypoint(waypoint: Waypoint) -> Self {
+        Self {
+            latest_version: waypoint.version(),
+            latest_epoch_waypoint: waypoint,
+            verifier: VerifierType::Waypoint(waypoint),
+        }
+    }
+
+    /// Create an initial trusted state that will trust the first genesis
+    /// presented to it.
+    ///
+    /// WARNING: this is obviously unsafe, as a malicious peer could present any
+    /// arbitrary genesis and this TrustedState would gladly accept it.
+    // TODO(philiphayes/dmitrip): remove this when waypoints are completely
+    // integrated with client code.
+    #[allow(non_snake_case)]
+    pub fn new_trust_any_genesis_WARNING_UNSAFE() -> Self {
+        let placeholder_genesis_li =
+            LedgerInfo::genesis(*ACCUMULATOR_PLACEHOLDER_HASH, ValidatorSet::empty());
+        // This Waypoint is just a placeholder so we don't have to change the API
+        // once this function is removed.
+        let placeholder_wp = Waypoint::new(&placeholder_genesis_li)
+            .expect("Genesis is always an epoch-change ledger info so this will never fail");
+        Self {
+            latest_version: 0,
+            latest_epoch_waypoint: placeholder_wp,
+            verifier: VerifierType::TrustedVerifier(EpochInfo::empty()),
+        }
+    }
+
+    /// Create an initial trusted state from an epoch change ledger info.
+    pub fn from_epoch_change_ledger_info(epoch_change_li: LedgerInfo) -> Result<Self> {
+        // Generate the EpochInfo from the new validator set.
+        let epoch_info = EpochInfo {
+            epoch: epoch_change_li.epoch() + 1,
+            verifier: Arc::new(
```

Comment:
this probably should be an associated method in ledger_info, but not related to this pr